modules/standard/Types.chpl

```diff
@@ -711,4 +711,110 @@ inline proc _bor_id(type t) return 0:t;
 pragma "no doc"
 inline proc _bxor_id(type t) return 0:t;
 
+
+/* Return the number of fields in a class or record as a param.
+   The count of fields includes types and param fields.
+ */
+proc numFields(type t) param
+  return __primitive("num fields", t);
+
+/* Get the name of the `i`th field in a class or record.
+   `i` should be in 1..numFields(t).
+
+   :arg t: a class or record type
+   :arg i: which field to get the name of
+   :returns: the name of the field, as a param string
+ */
+proc getFieldName(type t, param i:int) param
+  return __primitive("field num to name", t, i);
+
+pragma "no doc"
+proc fieldNumToName(type t, param i:int) param {
+  compilerError("fieldNumToName deprecated. Use getFieldName");
+}
+
+/* Get the `i`th field in a class or record.
+   `i` should be in 1..numFields(t).
+
+   :arg x: a class or record
+   :arg i: which field to get
+   :returns: an rvalue referring to that field.
+ */
+inline
+proc getField(x:?t, param i:int)
+  return __primitive("field value by num", x, i);
+
+// TODO ref return version of getField(x, int)
```

Comment:
Yes. The ref return intent is somehow causing the value of the param argument i to no longer be known at compile time in the usual way. Around  functionResolution.cpp:5583, in handling PRIM_FIELD_VALUE_BY_NUM, var is NULL. Is there some other way we should be getting the value of a param arg at compile time?

I thought I would put it off since I'm about to change ref return intent handling...